qsiprep/workflows/dwi/merge.py

```diff
@@ -167,21 +181,43 @@ def init_merge_and_denoise_wf(
     denoising_wfs = []
 
     # Get a data frame of the raw_dwi_files and their imaging parameters:
-    dwi_df = get_acq_parameters_df(raw_dwi_files)
+    dwi_df = get_acq_parameters_df(raw_dwi_files, layout=layout)
     for dwi_num, row in dwi_df.iterrows():
         dwi_num += 1  # start at 1
         dwi_file = row.BIDSFile
 
         # Conform each image to the requested orientation
         conformers.append(
-            pe.Node(ConformDwi(orientation=orientation), name="conform_dwis%02d" % dwi_num)
+            pe.Node(ConformDwi(orientation=orientation), name="conform_dwis%02d" % dwi_num),
         )
         conformers[-1].inputs.dwi_file = dwi_file
 
         if denoise_before_combining:
             # Build the denoising workflow
             _, fname, _ = split_filename(dwi_file)
             wf_name = _get_wf_name(fname).replace("preproc", "denoise")
+
+            # Try to find associated phase file.
+            # BIDSLayout.get_nearest() doesn't work well for this.
+            ents = layout.get_file(dwi_file).get_entities()
+            ents["part"] = "phase"
+            phase_files = layout.get(**ents)
+            phase_available = False
+            if len(phase_files) == 1:
```

Comment:
If the user hasn't validated their dataset is it possible they'll have more than 1?

Reply:
I suppose it's possible if there are other DWI scans that have a superset of this scan's entities. Something like:

```
sub-01_part-mag_dwi.nii.gz
sub-01_part-phase_dwi.nii.gz
sub-01_run-01_part-mag_dwi.nii.gz
sub-01_run-01_part-phase_dwi.nii.gz
```

In that case, `layout.get` using `sub-01_part-mag_dwi.nii.gz` as the base might find both `sub-01_part-phase_dwi.nii.gz` and `sub-01_run-01_part-phase_dwi.nii.gz`. Not sure if that's a valid BIDS dataset, but that situation could happen.

I can make the `get` call more strict by explicitly setting all other entities to `Query.NONE`.

Reply:
Take a look at the version I just pushed. WDYT?

Reply:
this should be sufficiently strict!